### DISCUSSION TITLE
BaseSettings Method to Customize Sources Programmatically

### DISCUSSION THREAD

**Discussion by nwchandler** [pydantic/pydantic#3027] [Question]
Hello.

I've been experimenting with Pydantic's settings management recently and really like it. I have been considering writing a small library to allow pulling config settings from AWS Parameter Store and Secrets Manager, primarily as a small project for myself, but also thinking it could be useful to others potentially - at the very least, I could reuse it in several projects.

Today, as far as I understand, the mechanism for customizing sources requires defining a class method, `customise_sources` on the Config class in the settings class, which returns a tuple of `SettingsSourceCallable`. In my case, and I suspect others, the SettingsSourceCallable that I need to include has some runtime variation - for example, SSM configuration settings may vary based on the environment in which the application is deployed. To provide this customization, I have a function that takes some arguments and returns a callable that is customized based on the settings. But, then, I have to go back and update the Config class in my settings in order to use this.

The snippet below, hopefully, helps to indicate how I'm doing this:

```python
class AppConfig(BaseSettings):
    username: str
    password: str
    service_url: str
    port: Optional[int]
    test: Optional[List[int]]

    class Config:
        env_prefix = "app_"
        extra = "ignore"

if __name__ == "__main__":
    client: "Client" = boto3.client("ssm")
    ssm_config: "SettingsSourceCallable" = parameter_store_config_builder(
        client=client, path="/config/"
    )

    AppConfig.Config.customise_sources = (
        lambda init_settings, env_settings, file_secret_settings: (
            ssm_config,
            env_settings,
        )
    )

    conf = AppConfig()
    print(conf.json())
```

My question is, is there a more idiomatic way to accomplish this today with Pydantic? And if not, would there be value in adding a method to update sources just prior to instantiating the settings?

Thanks!

**Comment by nymous** [comment]
Hey @nwchandler!
I don't really have an answer to your questions but I wanted to share how I did my own BaseSettings custom source ^^
In [Pydantic-Vault](https://github.com/nymous/pydantic-vault) I use mostly environment variables, as the Vault CLI already gets its configuration from the environment. I also let users configure the behavior with fields in the `Config` class.
I didn't even think of making a function that returns a configured Callable, that's an interesting approach and I will see if I change my project ^^ The issue that I have with the configuration through the `Config` class is that there is no completion for the fields I read, whereas Pycharm can autocomplete the "native" Pydantic Config fields.

Unfortunately I'm afraid my approach with environment variables or static config values will not work for you, as you need to pass a boto3 client correctly configured... Maybe if your function takes the responsibility of building such client by itself, and only taking the credentials as parameters...

**Reply by nwchandler** [reply]
Hi, @nymous! Thank you for the response! I haven't had a good chance to use Vault "for real" (outside of some lab work and proofs of concept), but I like it and I like your idea of creating a custom source for it to integrate with Pydantic. I'll keep it in mind if/when I get a chance to use Vault on a project!

My inspiration for a programmatic way to set config sources on configuration objects actually came from .NET. AWS has an extension for integrating Parameter Store with the .NET configuration system, and I kind of like the approach. I was hoping there might be a way to do something comparable with Pydantic. (For reference, the package is [here](https://github.com/aws/aws-dotnet-extensions-configuration/).) More generally, I like the [.NET Options Pattern](https://docs.microsoft.com/en-us/dotnet/core/extensions/options) and wanted to see if I could approximate it.

I may be able to do something with creating the client inside the function. My thought with having the caller specify it is, it should make it a little simpler to do unit tests. A mock client could be passed in for testing, so that users could avoid accidentally reaching out to AWS for tests. It would also allow the users of the library to have complete control over the client, including any lesser-used features that they may require of the client.

I'll keep giving it some more thought. I think I can come up with something reasonable, and if there's not a good way to update configuration sources programmatically, it may just require that I document how to go about setting configuration settings in both "production" and in unit tests.
